refine custom operations. re: #474

jamesmoriarty opened the pull request:
Regarding: #474

jamesmoriarty commented:
@IvanGoncharov aka graphql overlord can you please review?

IvanGoncharov commented on src/index.js:
```diff
@@ -463,3 +487,19 @@ function sendResponse(response: $Response, type: string, data: string): void {
+
+    validationCache[hash] = op();
+
+    return validationCache[hash];
```
There is no protection against DoS attack with a unique query per request that contains a lot of errors. Since validation errors contain AST nodes you will keep the entire query string in memory.
```js
const badQuery = `
query badQuery${counter++} {
  a b c d e #...
  a1 b1 c1 d1 e1 #...
  a2 b2 c2 d2 e2 #...
  #...
}`;
```
Since query name is different every time it means you will allocate huge amount of memory per request until your server crash.

jamesmoriarty replied:
Hey @IvanGoncharov, Thanks for getting back to me so quickly. I really appreciate it. I wouldn't advise using an associative object in production but it was the simplest method signature without introducing something like different storage strategies. 

You can use something like https://github.com/isaacs/node-lru-cache e.g. https://repl.it/repls/MonstrousAjarArchive

```js
var LRU = require("lru-cache"),
p = console.log,
options = {max: 10},
cache = new LRU(options);

cache['test'] = 'Hello World';
p(cache['test']);
p(cache);
```

jamesmoriarty commented:
Hey @IvanGoncharov, the set cache property signature was a mistake. I've changed it to `set(key, value)` and `get(key`. I've changed the test to use [LruMap](http://www.collectionsjs.com/lru-map).

IvanGoncharov commented on src/index.js:
```diff
@@ -251,7 +264,24 @@ function graphqlHTTP(options: Options): Middleware {
+        const string =
+          JSON.stringify(schema) +
+          JSON.stringify(source) +
+          validationRules.toString();
```
There is a problem with your caching approach `JSON.stringify(schema)` only stringify type names and not things like fields, enum values, etc. It also doesn't take into account user-provided callbacks that affect validation e.g. `parseLiteral`
https://github.com/graphql/graphql-js/blob/8c96dc8276f2de27b8af9ffbd71a4597d483523f/src/validation/rules/ValuesOfCorrectType.js#L186

Hashing `validationRules` is affected by the exact same problems.

I think that it's impossible to implement secure & effective validation caching for all use cases.
For example, in most cases, you schema and validation rules stay the same for all requests so converting them to string and hashing is an overhead.

I think the simplest and the most flexible solution would be to allow wrapping/replacing `validate` function, similar to #391 proposal but for `validate`.

@jamesmoriarty What do you think?

jamesmoriarty replied:
I think you're right @IvanGoncharov. We were concerned we had unsupported edge cases. We're only using a subset of graphql's feature so nothing was obviously broken. I'll see if I can invert the cache implementation with the wrapper function.

ivome commented:
I also think wrapping / replacing the validate function would be better. I wrote a validation rule to analyze the query complexity and reject complex queries that depends on the input variables: [graphql-query-complexity](https://github.com/slicknode/graphql-query-complexity)
To use something like this with validation caching we would have to implement some custom logic. 

Maybe we could even add a plugin system to `express-graphql` to have a well defined interface for things like this and even more complex extensions? This could also be used for #391. 

Initialization could look something like this: 

```javascript
app.get('/graphql', graphqlHTTP({
  schema: MyGraphQLSchema,
  graphiql: true,
  plugins: [
    new ValidationCachePlugin(options),
    // Other plugins
    new FakeResultPlugin(),

    // We could also add hooks for things like...
    new ProfilingPlugin(),
    // etc.
  ]
}));
```

Might be overkill for just this one feature and this would definitely need some more detailed conception but I just wanted to throw this out here to see what you think...

IvanGoncharov commented on src/index.js:
```diff
@@ -87,15 +86,14 @@ export type OptionsData = {
-    ) => $ReadOnlyArray<GraphQLError>,
-    get: (key: string) => $ReadOnlyArray<GraphQLError>,
-  |},
+  validationWrapper?: ?(
```
@jamesmoriarty Thanks for the effort 👍 

> Unsure about the best method signature. Should it pass in the response? Would you want the same for execute?

Looking more into the source code I notice we already use the same approach for `formatError`:
https://github.com/graphql/express-graphql/blob/7ab2bc958a2184fe35fb6cd137f369b5b3877dec/src/index.js#L75-L80
https://github.com/graphql/express-graphql/blob/7ab2bc958a2184fe35fb6cd137f369b5b3877dec/src/index.js#L336-L338

I think we should duplicate this approach for both `validate` and `execute`.
Since `graphql` is a peer dependency you need to add it into your `package.json` anyway so it mean you can just write:
```js
const { validate } = require('graphql');

app.post('/graphql', graphqlHTTP({
  schema: MyGraphQLSchema,
  graphiql: false
  validate(schema, documentAST, validationRules) {
    // some caching logic
    cache[hash] = validate(schema, documentAST, validationRules);
    // more caching logic
    return cache[hash]
  },
}));
```

jamesmoriarty replied:
Looks great. Will refactor towards this.

jamesmoriarty commented:
Thanks for the feedback @IvanGoncharov. Using the same pattern as the `formatError`worked great.

IvanGoncharov commented:
@jamesmoriarty Thanks 👍 
Can you also cleanup test cases (it doesn't make sense to introduce new dependencies just for tests) and make it more like `formatError`/#391?

jamesmoriarty commented:
Removed unnecessary dependencies and cleaned up the tests. Looks consistent with formatError and your execute PR.

<img width="785" alt="Screen Shot 2019-04-03 at 9 58 31 pm" src="https://user-images.githubusercontent.com/96487/55473964-b1696d80-565b-11e9-8cec-ac246da0da48.png">

IvanGoncharov commented:
@jamesmoriarty Thanks a lot 👍 I will try to merge it in the next few days.

One issue I still have is with both `validate` and `execute` is that they are not very intuitive names for the functionality they enable.
Maybe `replaceValidate`/`replaceExecute`?
Would be great to hear your thoughts?

jamesmoriarty commented:
I agree. The existing implementation made it possible. I've had a look at the graphql-js source to get ideas on consistency. I'm thinking `customValidateFn` `customExecuteFn` and possible deprecate `formatError` and change to `customFormatErrorFn`. What do you think @IvanGoncharov?

Based on:

<img width="720" alt="Screen Shot 2019-04-04 at 9 21 56 am" src="https://user-images.githubusercontent.com/96487/55517218-6bdc8d00-56bb-11e9-87ba-054b45add103.png">

IvanGoncharov commented:
@jamesmoriarty Great suggestion 👍 I like it.
I will update #391 to use `customExecuteFn`.

> I'm thinking customValidateFn customExecuteFn and possible deprecate formatError and change to customFormatErrorFn

Make sense 👍
I think we should support deprecated `formatError` until `1.0.0` release.
Do you want to work on PR for this?

jamesmoriarty commented:
Sure thing.

jamesmoriarty commented:
Made the change to the naming, deprecating `formatError` and updating the README. I imagine there is a better way to notify the user with the deprecation warning than using `console.error`. Any feedback appreciated.

jamesmoriarty commented on src/index.js:
```diff
@@ -380,7 +401,14 @@ function graphqlHTTP(options: Options): Middleware {
+        }
+
+        formatErrorFn =
+          optionsData.customFormatErrorFn || optionsData.formatError;
```
@IvanGoncharov interestingly I couldn't change this to:

```
optionsData.customFormatErrorFn || optionsData.formatError || formatError
// ...
.map(errorFormatFn) 
```

Without breaking the tests. It appears the 

```
.map(errorFormatFn || formatError) 
```

may be called before the options resolve. maybe around schema validation. 

Might be worth resolving outside this change.

IvanGoncharov commented on README.md:
```diff
@@ -91,6 +87,13 @@ The `graphqlHTTP` function accepts the following options:
+
+  * **`customFormatErrorFn`**: An optional function which will be used to format any
+    errors produced by fulfilling a GraphQL operation. If no function is
+    provided, GraphQL's default spec-compliant [`formatError`][] function will be used.
```
It should mention deprecated `formatError` and say it will be removed in `1.0.0`.

jamesmoriarty replied:
Good point. I've added below.

jamesmoriarty commented:
Added to README. Thanks for spending time on this @IvanGoncharov. Much appreciated.

IvanGoncharov commented:
@jamesmoriarty I made a few small fixes like replacing `console.error` with `console.warn` and fixing ESLint warnings.
